src/hmem.c

```diff
@@ -528,12 +528,59 @@ bool ofi_hmem_is_initialized(enum fi_hmem_iface iface)
 	return hmem_ops[iface].initialized;
 }
 
+void ofi_hmem_set_iface_filter(const char* iface_filter_str, bool* filter)
+{
+	int iface;
+	char* entry = NULL;
+	const char* token = ";";
+	const char* iface_labels[ARRAY_SIZE(hmem_ops)] = {
+		"system",
+		"cuda",
+		"rocr",
+		"ze",
+		"neuron",
+		"synapseai"
+	};
+
+	memset(filter, false, sizeof(bool) * ARRAY_SIZE(hmem_ops));
+
+	/* always enable system hmem interface */
+	filter[FI_HMEM_SYSTEM] = true;
+
+	entry = strtok(iface_filter_str, token);
+	while (entry != NULL) {
+		for (iface = 0; iface < ARRAY_SIZE(hmem_ops); iface++) {
+			if (!strcasecmp(iface_labels[iface], entry)) {
+				filter[iface] = true;
+				break;
+			}
+		}
+
+		entry = strtok(NULL, token);
+	}
+}
+
 void ofi_hmem_init(void)
 {
 	int iface, ret;
 	int disable_p2p = 0;
+	char* hmem_filter = NULL;
+	bool filter_hmem_ifaces = false;
+	bool iface_filter_array[ARRAY_SIZE(hmem_ops)];
+
+	fi_param_define(NULL, "hmem", FI_PARAM_STRING,
+			"List of hmem interfaces to attempt to initialize (default: all available interfaces)");
+	fi_param_get_str(NULL, "hmem", &hmem_filter);
+
+	if (hmem_filter && strlen(hmem_filter) != 0) {
+		ofi_hmem_set_iface_filter(hmem_filter, &iface_filter_array);
+		filter_hmem_ifaces = true;
+	}
```

Comment:
A question for discussion: does it make sense to treat empty setting of this variable (i.e. `FI_HMEM=`) as disabling all the HMEM interface?

Reply:
I've been thinking about it as well. FI_HMEM set to the empty string could be used to say 'do not initialize anything' or just enable FI_HMEM_SYSTEM.

I think an empty setting should be treated as FI_HMEM_SYSTEM. I cannot conceive of a use case where FI_HMEM_SYSTEM shouldn't be available.

Reply:
It seems like FI_HMEM="" should be treated as a user error and a warn message logged. Since FI_HMEM="" is ambiguous, rather than make the assumption that the user wants FI_HMEM_SYSTEM, I think we should init all HMEM ifaces. Yes, this will result in more dlopens. But, the user needs to use this env properly if they want to avoid this.

Reply:
FI_HMEM="" is currently set up to enable all provider. However, after implementing it like that, I saw it as a case for users to say -- do not init anything. 

The only question in my mind is whether we should make it very clear -- and init nothing, init everything, or throw an error. 

I would lean toward throwing an error perhaps because if a user writes a script to launch their application, and makes a mistake -- then they could fill out FI_HMEM with an empty string. It could be useful to them to log that as an unexpected input.

Reply:
I think giving a message with FI_WARN() should be sufficient.

Reply:
Agreed. That is what I meant, but I can see where what I said might have been confusing.

Reply:
```
C02ZH00YLVCJ:libfabric jswaro$ FI_LOG_LEVEL=warn FI_HMEM="" ./util/fi_info -p udp
libfabric:69111:1698431409::core:core:ofi_hmem_init():587<warn> zero-length FI_HMEM provided, enabling all interfaces
fi_getinfo: -96 (No message available on STREAM)
C02ZH00YLVCJ:libfabric jswaro$ FI_LOG_LEVEL=warn FI_HMEM="something" ./util/fi_info -p udp
libfabric:69125:1698431414::core:core:ofi_hmem_set_iface_filter():562<warn> unknown HMEM interface specified in FI_HMEM, entry="something"
fi_getinfo: -96 (No message available on STREAM)
C02ZH00YLVCJ:libfabric jswaro$ FI_LOG_LEVEL=warn FI_HMEM="cuda" ./util/fi_info -p udp
fi_getinfo: -96 (No message available on STREAM)
C02ZH00YLVCJ:libfabric jswaro$  ./util/fi_info -p udp
fi_getinfo: -96 (No message available on STREAM)
```

Examples of each case above.

Reply:
```
C02ZH00YLVCJ:libfabric jswaro$ FI_LOG_LEVEL=warn FI_HMEM="cuda;rocr;something" ./util/fi_info -p udp
libfabric:69209:1698431519::core:core:ofi_hmem_set_iface_filter():562<warn> unknown HMEM interface specified in FI_HMEM, entry="something"
fi_getinfo: -96 (No message available on STREAM)
```